### PR TITLE
Port four bar linkage and differential transmission loaders from ROS1

### PR DESCRIPTION
**Pull request by VX792**
Plus minor changes to parsing
Should I create test cases with invalid roles, since they were present in the ROS1 version?

Issue: #599

### CODE REVIEW

**Review by bijoua29 — APPROVED** [review]

File: transmission_interface/test/differential_transmission_loader_test.cpp [hunk 1]
```diff
@@ -0,0 +1,469 @@
...
+  // Transmission loader
+  TransmissionPluginLoader loader;
+  std::shared_ptr<transmission_interface::TransmissionLoader> transmission_loader =
+    loader.create(infos[0].transmissions[0].type);
+  ASSERT_TRUE(nullptr != transmission_loader);
+
+  std::shared_ptr<transmission_interface::Transmission> transmission;
+  const hardware_interface::TransmissionInfo & info = infos[0].transmissions[0];
+  transmission = transmission_loader->load(info);
+  ASSERT_TRUE(nullptr != transmission);
```
Comment (destogl) [inline transmission_interface/test/differential_transmission_loader_test.cpp:108]: Please check best practices comment.

File: transmission_interface/test/differential_transmission_loader_test.cpp [hunk 2]
```diff
@@ -0,0 +1,469 @@
...
+
+  // Transmission loader
+  TransmissionPluginLoader loader;
+  std::shared_ptr<transmission_interface::TransmissionLoader> transmission_loader =
+    loader.create(infos[0].transmissions[0].type);
+  ASSERT_TRUE(nullptr != transmission_loader);
+
+  std::shared_ptr<transmission_interface::Transmission> transmission = nullptr;
+  const hardware_interface::TransmissionInfo & info = infos[0].transmissions[0];
+  transmission = transmission_loader->load(info);
```
Comment (destogl) [inline transmission_interface/test/differential_transmission_loader_test.cpp:180]: Aren't we throwing an exception here in `SimpleTransmission`? If so we should do the same.

**Review by destogl — COMMENTED** [review]
First set of comments.

**Review by bmagyar — CHANGES_REQUESTED** [review]

File: hardware_interface/include/hardware_interface/hardware_info.hpp [hunk 3]
```diff
@@ -84,6 +84,7 @@ struct ActuatorInfo
   std::string name;
   std::vector<std::string> interfaces;
   std::string role;
+  double mechanical_reduction = 1.0;
```
[the remark this answers is not in the document]
Comment (VX792) [reply hardware_interface/include/hardware_interface/hardware_info.hpp:87]: Ok, I'll do this and address the rest of your comments in my next commit

File: transmission_interface/test/differential_transmission_loader_test.cpp (hunk 1 above)
Comment (VX792) [reply transmission_interface/test/differential_transmission_loader_test.cpp:108]: The best practices comment says that either the comparison  ASSERT_TRUE(nullptr != transmission) is redundant, or there is possible null pointer dereference. I don't think there's a case where the transmission_loader would be a nullptr, so should I just leave out these comparisons?

File: transmission_interface/src/four_bar_linkage_transmission_loader.cpp [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+    const auto jnt_offset1 = transmission_info.joints.at(0).offset;
+    const auto jnt_offset2 = transmission_info.joints.at(1).offset;
+
+    std::shared_ptr<Transmission> transmission(new FourBarLinkageTransmission(
+      {act_reduction1, act_reduction2}, {jnt_reduction1, jnt_reduction2},
+      {jnt_offset1, jnt_offset2}));
+    return transmission;
+  }
+  catch (const Exception & ex)
```
[the remark this answers is not in the document]
Comment (VX792) [reply transmission_interface/src/four_bar_linkage_transmission_loader.cpp:48]: Yes, for example if we have one actuator and 2 joints in our urdf. I've just tested it and the test case fails with
C++ exception with description "vector::_M_range_check: __n (which is 1) >= this->size() (which is 1)" thrown in the test body.

The initial SizeIs checks passes, and in the load function we try to access the nonexistent 
transmission_info.actuators.at(1).mechanical_reduction; variable. Is it okay if I catch an std::exception type here?

File: transmission_interface/src/differential_transmission_loader.cpp [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    std::shared_ptr<Transmission> transmission(new DifferentialTransmission(
+      {act_reduction1, act_reduction2}, {jnt_reduction1, jnt_reduction2},
+      {jnt_offset1, jnt_offset2}));
+    return transmission;
+  }
+  catch (const std::exception & ex)
+  {
+    RCLCPP_ERROR(
+      rclcpp::get_logger("differential_transmission_loader"),
+      "Failed to construct transmission '%s'", ex.what());
```
Comment (bmagyar) [inline transmission_interface/src/differential_transmission_loader.cpp:52]: do we want `RLCPP_ERROR` or `std::cerr` here like the other classes? I believe `std::cerr` was used to stop a crash

**Review by bmagyar — APPROVED** [review]

**Pull request merged**